main/src/com/google/refine/model/changes/RowAdditionChange.java

```diff
@@ -0,0 +1,122 @@
+/*******************************************************************************
+ * Copyright (C) 2024, OpenRefine contributors
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+package com.google.refine.model.changes;
+
+import java.io.IOException;
+import java.io.LineNumberReader;
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+import com.google.refine.ProjectManager;
+import com.google.refine.history.Change;
+import com.google.refine.model.Project;
+import com.google.refine.model.Row;
+import com.google.refine.util.Pool;
+
+public class RowAdditionChange implements Change {
+
+    private final List<Row> _additionalRows;
+
+    private final int _insertionIndex;
+    private static String COUNT_FIELD = "count"; // Field name for sizes of `_additionalRows` in `change.txt`
+    private static String INDEX_FIELD = "index"; // Field name for index to insert new rows in `change.txt`
+    private static String EOC = "/ec/"; // end of change marker, not end of file, in `change.txt`
+
+    public RowAdditionChange(List<Row> additionalRows, int insertionIndex) {
```

Comment:
I wonder if it would make sense to only store the number of additional rows, since they are known to be blank. That would be simpler and would save memory.

On the other hand you could say that it's useful to make it possible to support non-blank rows at the change level for later, so that projects created with later versions of OpenRefine could also be read by earlier versions, but that's generally something that's too hard to support reliably to advertise it to users. Also, if we are to support arbitrary rows, we would probably need to serialize the recon pool too, so that rows can contain recon values.

No strong feelings either way.

Reply:
I have strong feelings in favor of the current implementation, sending the server rows that are blank. This design decision setups future development for supporting non-blank rows. I think adding rows of new data into an existing project is the primary task that this feature will address and padding a table with blank rows is an edge case, although [supporting expressive use of data tables is still important](https://www.tableau.com/research/publications/untidy-data-unreasonable-effectiveness-tables) and I understand the pragmatic importance of incremental enhancements in a widely used application such as OpenRefine.

I also think OpenRefine benefits from a new model change for a small subset of new rows, as per this TODO item in `MassRowChange`. `RowAdditionChange` begins to address it.
 
https://github.com/OpenRefine/OpenRefine/blob/183246f6b0fde965d4d2d7bdcf6595dc97e857a9/main/src/com/google/refine/model/changes/MassRowChange.java#L49-L50

Reply:
I'm OK with row data which is required to be blank as long we we're clear about that restriction in the code comments and that the tests match whatever we're claiming to support (I haven't reviewed the tests yet).